*** ../cil/src/formatparse.mly	2006-05-21 06:14:15.000000000 +0200
--- ../cil_patch/src/formatparse.mly	2006-06-20 16:22:57.000000000 +0200
***************
*** 1,3 ****
--- 1,5 ----
+ /* MODIF: Loop constructor replaced by 3 constructors: While, DoWhile, For. */
+ 
  /*(* Parser for constructing CIL from format strings *)
  (*
   *
***************
*** 1352,1357 ****
--- 1354,1360 ----
                              mkCast e !upointType
                            else e
                          in
+ (*
                          mkStmt 
                            (Loop (mkBlock [ mkStmt 
                                               (If(e,
***************
*** 1360,1366 ****
                                                               (Break loc) ],
                                                   loc));
                                             $5 mkTemp loc args ],
!                                  loc, None, None)))
                     } 
  |   instr_list    { (fun mkTemp loc args -> 
                         mkStmt (Instr ($1 loc args)))
--- 1363,1372 ----
                                                               (Break loc) ],
                                                   loc));
                                             $5 mkTemp loc args ],
!                                  loc, None, None))
! *)
! 			mkStmt
! 			  (While (e, mkBlock [ $5 mkTemp loc args ], loc)))
                     } 
  |   instr_list    { (fun mkTemp loc args -> 
                         mkStmt (Instr ($1 loc args)))
