--- ../cil/Makefile.in.orig	2008-12-31 19:08:43.000000000 +0100
+++ ../cil/Makefile.in	2008-12-31 19:09:00.000000000 +0100
@@ -212,7 +212,7 @@
 # build two libraries
 .PHONY: cillib libcil
 ifeq ($(NATIVECAML),1)
-cillib: $(OBJDIR)/cil.$(CMXA) $(OBJDIR)/libcil.a
+cillib: $(OBJDIR)/cil.$(CMXA) # $(OBJDIR)/libcil.a
 else
 cillib: $(OBJDIR)/cil.$(CMXA) 
 endif
@@ -243,9 +243,9 @@
 	echo "   Zrapp.feature;" >> $@  
 endif
 # Now the extra features, with the first letter capitalized
-	echo -ne \
+	echo \
           $(foreach f,@EXTRAFEATURES@, \
-             `echo $f | cut -c 1 | tr "[a-z]" "[A-Z]"``echo $f | cut -c 2-`".feature;\n") >> $@
+             `echo $f | cut -c 1 | tr "[a-z]" "[A-Z]"``echo $f | cut -c 2-`".feature;") >> $@
 	echo "]"   >>$@ 
 # Must delete main.d and remake it, because it may have been made
 # before feature_config existed.
