*** ../cil/src/ext/pta/ptranal.ml	2006-05-21 06:14:15.000000000 +0200
--- ../cil_patch/src/ext/pta/ptranal.ml	2006-06-21 11:55:25.414890423 +0200
***************
*** 1,3 ****
--- 1,5 ----
+ (* MODIF: Loop constructor replaced by 3 constructors: While, DoWhile, For. *)
+ 
  (*
   *
   * Copyright (c) 2001-2002,
***************
*** 312,318 ****
--- 314,328 ----
      | Switch (e, b, sl, l) ->
          analyze_block b;
          List.iter analyze_stmt sl
+ (*
      | Loop (b, l, _, _) -> analyze_block b
+ *)
+     | While (_, b, _) -> analyze_block b
+     | DoWhile (_, b, _) -> analyze_block b
+     | For (bInit, _, bIter, b, _) ->
+ 	analyze_block bInit;
+ 	analyze_block bIter;
+ 	analyze_block b
      | Block b -> analyze_block b
      | TryFinally (b, h, _) ->
          analyze_block b;
