*** kildall.ml.orig	2009-08-16 15:45:21.000000000 +0200
--- kildall.ml	2009-08-16 15:45:27.000000000 +0200
***************
*** 252,259 ****
    
    (** val basic_block_map : positive list PTree.t -> positive -> bbmap **)
    
!   let basic_block_map successors entrypoint x =
!     is_basic_block_head entrypoint (make_predecessors successors) x
    
    (** val basic_block_list :
        positive list PTree.t -> bbmap -> positive list **)
--- 252,259 ----
    
    (** val basic_block_map : positive list PTree.t -> positive -> bbmap **)
    
!   let basic_block_map successors entrypoint =
!     is_basic_block_head entrypoint (make_predecessors successors)
    
    (** val basic_block_list :
        positive list PTree.t -> bbmap -> positive list **)
